contracts/verifier.sol

```diff
@@ -0,0 +1,260 @@
+        _;
+    }
+
+    // @TODO(shrugs) - this allows anyone to challenge an empty claim
+    //  is this what we want?
+    modifier onlyChallenger(uint sessionId) {
+        var session = sessions[sessionId];
+        if (session.challenger == address(0)) {
+            session.challenger = msg.sender;
+        } else {
+            require(msg.sender == session.challenger);
+        }
+        _;
+    }
+
+    function query(uint sessionId, uint step)
+        onlyChallenger(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+
+        bool isFirstStep = s.medStep == 0;
+        bool haveMedHash = s.medHash != bytes32(0);
+        assert(isFirstStep || haveMedHash);
+        // ^ invariant if the step has been set but we don't have a hash for it
+
+        if (step == s.lowStep && step + 1 == s.medStep) {
+            // final step of the binary search (lower end)
+            s.highHash = s.medHash;
+            s.highStep = step + 1;
+        } else if (step == s.medStep && step + 1 == s.highStep) {
+            // final step of the binary search (upper end)
+            s.lowHash = s.medHash;
+            s.lowStep = step;
+        } else {
+            // this next step must be in the correct range
+            require(step > s.lowStep && step < s.highStep);
+
+            // if this is NOT the first query, update the steps and assign the correct hash
+            // (if this IS the first query, we just want to initialize medStep and medHash)
+            if (!isFirstStep) {
+                if (step < s.medStep) {
+                    // if we're iterating lower,
+                    //   the new highest is the current middle
+                    s.highStep = s.medStep;
+                    s.highHash = s.medHash;
+                } else if (step > s.medStep) {
+                    // if we're iterating upwards,
+                    //   the new lowest is the current middle
+                    s.lowStep = s.medStep;
+                    s.lowHash = s.medHash;
+                } else {
+                    // and if we're requesting the midStep that we've already requested,
+                    //   there's nothing to do.
+                    // @TODO(shrugs) - should this revert?
+                }
+            }
+
+            s.medStep = step;
+            s.medHash = bytes32(0);
+        }
+        s.lastChallengerMessage = now;
+        NewQuery(sessionId);
+    }
+
+    function respond(uint sessionId, uint step, bytes32 hash)
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        // Require step to avoid replay problems
+        require(step == s.medStep);
+        if (hash == 0) {
+            // Special "fold" signal
+            // @TODO(shrugs) - is this correct? should the claimant not fold here?
+            challengerConvicted(sessionId);
+            return;
+        }
+
+        // record the claimed hash
```

Comment:
Add `require(s.medHash == 0)` so that the claimant cannot call this multiple times.

Reply:
Added in b59f0b45ec170a09c9d9f526239996c853d3878c